core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java

```diff
@@ -85,6 +95,29 @@ public void deleteFile(String location) {
     io(location).deleteFile(location);
   }
 
+  @Override
+  public void deleteFiles(Iterable<String> pathsToDelete) throws BulkDeletionFailureException {
+    // peek at the first element to determine the type of FileIO
+    Iterator<String> originalIterator = pathsToDelete.iterator();
+    if (!originalIterator.hasNext()) {
+      return;
+    }
+
+    PeekingIterator<String> iterator = Iterators.peekingIterator(originalIterator);
+    SupportsBulkOperations fileIO = io(iterator.peek());
+    fileIO.deleteFiles(() -> iterator);
```

Comment:
Did we resolve this? I think we probably want to pass the iterable here rather than the iterator to make sure we close everything properly

Reply:
No not yet, I'm still thinking through the best solution. Passing the original iterable could cause an additional request to be made, i.e. one additional request for the peek, so I was hoping to avoid that.